src/colmap/estimators/covariance.cc

```diff
@@ -31,741 +31,400 @@
-}
-
-bool BundleAdjustmentCovarianceEstimatorBase::HasValidFullCovariance() const {
-  return cov_variables_.size() != 0;
-}
-
-double BundleAdjustmentCovarianceEstimatorBase::GetPoseCovarianceByIndex(
-    int row, int col) const {
-  THROW_CHECK(HasValidPoseCovariance());
-  return cov_poses_(row, col);
-}
-
-Eigen::MatrixXd
-BundleAdjustmentCovarianceEstimatorBase::GetPoseCovarianceBlockOperation(
-    int row_start,
-    int col_start,
-    int row_block_size,
-    int col_block_size) const {
-  THROW_CHECK(HasValidPoseCovariance());
-  return cov_poses_.block(row_start, col_start, row_block_size, col_block_size);
-}
-
-double BundleAdjustmentCovarianceEstimatorBase::GetCovarianceByIndex(
-    int row, int col) const {
-  THROW_CHECK(HasValidFullCovariance());
-  return cov_variables_(row, col);
-}
-
-Eigen::MatrixXd
-BundleAdjustmentCovarianceEstimatorBase::GetCovarianceBlockOperation(
-    int row_start,
-    int col_start,
-    int row_block_size,
-    int col_block_size) const {
-  THROW_CHECK(HasValidFullCovariance());
-  return cov_variables_.block(
-      row_start, col_start, row_block_size, col_block_size);
-}
-
-Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance()
-    const {
-  THROW_CHECK(HasValidPoseCovariance());
-  return cov_poses_;
-}
-
-Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance(
-    image_t image_id) const {
-  THROW_CHECK(HasReconstruction());
-  THROW_CHECK(HasPose(image_id));
-  int index = GetPoseIndex(image_id);
-  int num_params_pose = GetPoseTangentSize(image_id);
-  return GetPoseCovarianceBlockOperation(
-      index, index, num_params_pose, num_params_pose);
-}
-
-Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance(
-    const std::vector<image_t>& image_ids) const {
-  THROW_CHECK(HasReconstruction());
-  std::vector<int> indices;
-  for (const auto& image_id : image_ids) {
-    THROW_CHECK(HasPose(image_id));
-    int index = GetPoseIndex(image_id);
-    int num_params_pose = GetPoseTangentSize(image_id);
-    for (int i = 0; i < num_params_pose; ++i) {
-      indices.push_back(index + i);
+  // Notice that here "a" refers to the block of pose + other parameters.
+  const Eigen::SparseMatrix<double> J_a =
+      J_full.block(0, 0, J_full.rows(), J_full.cols() - point_num_params);
+  const Eigen::SparseMatrix<double> J_p = J_full.block(
+      0, J_full.cols() - point_num_params, J_full.rows(), point_num_params);
+  const Eigen::SparseMatrix<double> H_aa = J_a.transpose() * J_a;
+  const Eigen::SparseMatrix<double> H_ap = J_a.transpose() * J_p;
+  const Eigen::SparseMatrix<double> H_pa = H_ap.transpose();
+  Eigen::SparseMatrix<double> H_pp = J_p.transpose() * J_p;
+  // In-place computation of H_pp_inv.
+  Eigen::SparseMatrix<double>& H_pp_inv = H_pp;
+  int point_param_idx = 0;
+  for (const detail::PointParam& point : points) {
+    const Eigen::Matrix3d H_pp_idx =
+        H_pp.block(point_param_idx, point_param_idx, 3, 3) +
```

Comment:
We should use the corresponding parameter size instead of 3 here. This will break if the points are partially fixed, which is actually practical if one has a ground control points with x and y.